Zend/zend_enum.c

```diff
@@ -0,0 +1,216 @@
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__tostring, "__toString");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__debugInfo, "__debugInfo");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__serialize, "__serialize");
+	ZEND_ENUM_DISALLOW_MAGIC_METHOD(__unserialize, "__unserialize");
+
+	char* forbidden_methods[] = {
+		"__sleep",
+		"__wakeup",
+		"__set_state",
+	};
+
+	uint32_t forbidden_methods_length = sizeof(forbidden_methods) / sizeof(forbidden_methods[0]);
+	for (uint32_t i = 0; i < forbidden_methods_length; ++i) {
+		char *forbidden_method = forbidden_methods[i];
+
+		if (zend_hash_str_find_ptr(&ce->function_table, forbidden_method, strlen(forbidden_method))) {
+			zend_error_noreturn(E_COMPILE_ERROR, "Enum may not include magic method %s", forbidden_method);
+		}
+	}
+}
+
+void zend_verify_enum(zend_class_entry *ce)
+{
+	zend_verify_enum_properties(ce);
+	zend_verify_enum_magic_methods(ce);
+}
+
+ZEND_COLD zval *zend_enum_write_property(zend_object *object, zend_string *member, zval *value, void **cache_slot)
+{
+	ZEND_ENUM_PROPERTY_ERROR();
+	return &EG(uninitialized_zval);
+}
+
+static ZEND_COLD void zend_enum_unset_property(zend_object *object, zend_string *member, void **cache_slot) /* {{{ */
+{
+	ZEND_ENUM_PROPERTY_ERROR();
+}
+
+void zend_register_enum_ce(void)
+{
+	memcpy(&enum_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	enum_handlers.write_property = zend_enum_write_property;
+	enum_handlers.unset_property = zend_enum_unset_property;
+}
+
+ZEND_NAMED_FUNCTION(zend_enum_cases_func)
+{
+	zend_class_entry *ce = execute_data->func->common.scope;
+	zend_class_constant *c;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	array_init(return_value);
+
+	ZEND_HASH_FOREACH_PTR(&ce->constants_table, c) {
+		if (!(c->const_flags & ZEND_CLASS_CONST_IS_CASE)) {
+			continue;
+		}
+		zval *zv = &c->value;
+		if (Z_TYPE_P(zv) == IS_CONSTANT_AST) {
+			zval_update_constant_ex(zv, c->ce);
+			if (UNEXPECTED(EG(exception) != NULL)) {
+				return;
+			}
+		}
+		if (ce->enum_scalar_type == IS_UNDEF) {
+			zend_hash_next_index_insert(Z_ARRVAL_P(return_value), zv);
+		} else {
+			zend_object *zobj = Z_OBJ_P(zv);
+			zend_string *value_string = zend_string_init("value", strlen("value"), 0);
+			zval tmp;
+			zval *value_zv = zend_read_property_ex(ce, zobj, value_string, 0, &tmp);
+			zend_string_free(value_string);
+
+			if (ce->enum_scalar_type == IS_LONG) {
+				zend_hash_index_update(Z_ARRVAL_P(return_value), Z_LVAL_P(value_zv), zv);
+			} else {
+				ZEND_ASSERT(ce->enum_scalar_type == IS_STRING);
+				zend_symtable_update(Z_ARRVAL_P(return_value), Z_STR_P(value_zv), zv);
+			}
+		}
+	} ZEND_HASH_FOREACH_END();
+}
+
+ZEND_NAMED_FUNCTION(zend_enum_from_func)
+{
+	zend_class_entry *ce = execute_data->func->common.scope;
+	zend_string *string_key;
+	zend_long long_key;
+
+	zval *case_name_zv;
+	if (ce->enum_scalar_type == IS_LONG) {
+		if (zend_parse_parameters(ZEND_NUM_ARGS(), "l", &long_key) == FAILURE) {
```

Comment:
Only allowing l/S for int or string enums respectively technically violates the method signature in the interface. Is that a problem?

Reply:
This was specified accordingly.